### PR TITLE
Add min_count and max_count options to ec2 module

### PR DESCRIPTION
**Pull request by wrouesnel**
##### SUMMARY
These two options fulfill the same purpose as the exact_count option, but allow
for loosely specifying the number of instances or bounding it.

min_count ensures at least a certain number of instances are running, max_count
ensures no more then a certain number are running.

min_count and max_count can be used together, but are mutually exclusive to
exact_count (which is functionally identical to setting both values to the same
number).

These options make performing non Amazon autoscaling much easier, and
in particular allow idempotent deploy and bootstrapping ("ensure at least 1
production machine exists" can now be written into a play.)

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
ec2

##### ANSIBLE VERSION
```
ansible 2.4.0 (ansible-modules-core/pull/2240 7a9ce1b502) last updated 2017/06/17 12:08:27 (GMT +1100)
  config file = None
  configured module search path = ['/home/will/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/will/src/ansible-work/ansible/lib/ansible
  executable location = /usr/local/bin/ansible
  python version = 3.5.3 (default, Jan 19 2017, 14:11:04) [GCC 6.3.0 20170118]
```

### CODE REVIEW

**Comment by wrouesnel** [comment]
Migrated from https://github.com/ansible/ansible-modules-core/pull/2240

File: lib/ansible/modules/cloud/amazon/ec2.py
Comment (mattclay) [inline lib/ansible/modules/cloud/amazon/ec2.py]: @wrouesnel CI failure due to PEP 8 issue:

```
2017-01-28 07:26:38 ERROR: PEP 8: lib/ansible/modules/cloud/amazon/ec2.py:930:1: W293 blank line contains whitespace (legacy)
```

**Comment by wrouesnel** [comment]
Rebased and updated.

**Comment by wrouesnel** [comment]
Rebased and updated again.

**Comment by wrouesnel** [comment]
@mattclay Any chance of getting this looked at again? AFAIK the functionality here is still not available in the ec2 module and I believe valuable. It'd be good to get it merged.

File: lib/ansible/modules/cloud/amazon/ec2.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2.py]: Can you remove these and only import specifically what is needed with the other imports?

File: lib/ansible/modules/cloud/amazon/ec2.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2.py]: I don't see this being used, can you remove it?

File: lib/ansible/modules/cloud/amazon/ec2.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2.py]: Can you put that back rather than importing everything from ansible.module_utils.basic?

File: lib/ansible/modules/cloud/amazon/ec2.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2.py]: Can delete this line

**Review by s-hertel — CHANGES_REQUESTED** [review]

**Comment by s-hertel** [comment]
@wrouesnel I'll test this more thoroughly within a few days. This was just a cursory review.

File: lib/ansible/modules/cloud/amazon/ec2.py
Comment (wrouesnel) [reply lib/ansible/modules/cloud/amazon/ec2.py]: Looks like a bad rebase at some point. Fixed.

File: lib/ansible/modules/cloud/amazon/ec2.py
Comment (wrouesnel) [reply lib/ansible/modules/cloud/amazon/ec2.py]: This one crept in because at some point a whole bunch of modules were doing it and I was attempting to follow the style. Fixed.

File: lib/ansible/modules/cloud/amazon/ec2.py
Comment (wrouesnel) [reply lib/ansible/modules/cloud/amazon/ec2.py]: Me neither - another one which I believe I inherited from a rebase at some point. Fixed.

File: lib/ansible/modules/cloud/amazon/ec2.py
Comment (wrouesnel) [reply lib/ansible/modules/cloud/amazon/ec2.py]: Fixed.

File: lib/ansible/modules/cloud/amazon/ec2.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2.py]: min_count =

File: lib/ansible/modules/cloud/amazon/ec2.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2.py]: if min_count <=

File: lib/ansible/modules/cloud/amazon/ec2.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2.py:1645]: Could you add a check somewhere that min_count <= max_count?

File: lib/ansible/modules/cloud/amazon/ec2.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2.py]: Can you put the whitespace back, please?

File: lib/ansible/modules/cloud/amazon/ec2.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2.py:1706]: You could change this to `if not (module.params.get('exact_count') and module.params.get('min_count') and module.params.get('max_count')):`

File: lib/ansible/modules/cloud/amazon/ec2.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2.py]: Can you add an example using these?

**Review by s-hertel — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/amazon/ec2.py
Comment (wrouesnel) [reply lib/ansible/modules/cloud/amazon/ec2.py]: Embarrassing ... (and fixed).

File: lib/ansible/modules/cloud/amazon/ec2.py
Comment (wrouesnel) [reply lib/ansible/modules/cloud/amazon/ec2.py]: Fixed.

File: lib/ansible/modules/cloud/amazon/ec2.py
Comment (wrouesnel) [reply lib/ansible/modules/cloud/amazon/ec2.py:1706]: I'd somewhat prefer not to because while it's terser, IMO it's harder to read what's going on in.

File: lib/ansible/modules/cloud/amazon/ec2.py
Comment (wrouesnel) [reply lib/ansible/modules/cloud/amazon/ec2.py]: Fixed.

**Review by s-hertel — APPROVED** [review]

**Comment by s-hertel** [comment]
shipit

**Comment by wrouesnel** [comment]
Any progress on merging this ?

**Comment by s-hertel** [comment]
@ryansb @willthames @michael-dev2rights Can I get another review on this, please, before merging?

**Review by willthames — APPROVED** [review]

**Comment by willthames** [comment]
lgtm @s-hertel

**Comment by ryansb** [comment]
I feel like this is drifting out of the intended scope for this module: for cases where you need "between X and Y instances" I'd think an autoscaling group would be the way to go, since it offers termination policies and other niceties the ec2 module will never be able to match.

**Pull request closed without merging**
